src/statefulmodelinstance.cpp

```diff
@@ -180,117 +180,84 @@ const Status StatefulModelInstance::validateSpecialKeys(const tensorflow::servin
-    status = performInference(inferRequest);
-    timer.stop(PREDICTION);
-    if (!status.ok())
-        return status;
-    SPDLOG_DEBUG("Prediction duration in model {}, version {}, nireq {}: {:.3f} ms",
-        requestProto->model_spec().name(), getVersion(), executingInferId, timer.elapsed<microseconds>(PREDICTION) / 1000);
-
-    timer.start(SERIALIZE);
-    OutputGetter<ov::InferRequest&> outputGetter(inferRequest);
-    status = serializePredictResponse(outputGetter, getOutputsInfo(), responseProto, getTensorInfoName);
-    timer.stop(SERIALIZE);
-    if (!status.ok())
-        return status;
-    SPDLOG_DEBUG("Serialization duration in model {}, version {}, nireq {}: {:.3f} ms",
-        requestProto->model_spec().name(), getVersion(), executingInferId, timer.elapsed<microseconds>(SERIALIZE) / 1000);
-
-    timer.start(POSTPROCESS);
-    status = postInferenceProcessing(responseProto, inferRequest, sequence, sequenceProcessingSpec);
-    timer.stop(POSTPROCESS);
-    if (!status.ok())
-        return status;
-    SPDLOG_DEBUG("Postprocessing duration in model {}, version {}, nireq {}: {:.3f} ms",
-        requestProto->model_spec().name(), getVersion(), executingInferId, timer.elapsed<microseconds>(POSTPROCESS) / 1000);
-
-    sequenceLock.unlock();
+    sequenceLock = std::make_unique<std::unique_lock<std::mutex>>(sequence->getMutex());
+    sequenceManagerLock->unlock();
+    return StatusCode::OK;
+}
+template <>
+Status StatefulRequestProcessor<tensorflow::serving::PredictRequest, tensorflow::serving::PredictResponse>::preInferenceProcessing(ov::InferRequest& inferRequest) {
+    if (sequenceProcessingSpec.getSequenceControlInput() == SEQUENCE_START) {
+        // On SEQUENCE_START reset memory state of infer request to default
+        for (auto&& state : inferRequest.query_state()) {
+            state.reset();
+        }
+    } else {
+        // For next requests in the sequence set infer request memory state to the last state saved by the sequence
+        const sequence_memory_state_t& sequenceMemoryState = sequence->getMemoryState();
+        for (auto&& state : inferRequest.query_state()) {
+            auto stateName = state.get_name();
+            if (!sequenceMemoryState.count(stateName))
+                return StatusCode::INTERNAL_ERROR;
+            state.set_state(sequenceMemoryState.at(stateName));
+        }
+    }
+    return StatusCode::OK;
+}
+template <>
+Status StatefulRequestProcessor<tensorflow::serving::PredictRequest, tensorflow::serving::PredictResponse>::postInferenceProcessing(tensorflow::serving::PredictResponse* response, ov::InferRequest& inferRequest) {
+    // Reset inferRequest states on SEQUENCE_END
     if (sequenceProcessingSpec.getSequenceControlInput() == SEQUENCE_END) {
-        sequenceManagerLock.lock();
-        status = sequenceManager->removeSequence(sequenceId);
+        SPDLOG_DEBUG("Received SEQUENCE_END signal. Reseting model state and removing sequence");
```

Comment:
Why are we removing clearing the sequence in sequence manager? 
If that was needed, we should update log not to include "and removing sequence" part

Reply:
It is just moved to tatefulRequestProcessor<T>::release().
I will split the log to indicate that.